### PR TITLE
[feat] Support prompt level dynamic 

### PR DESCRIPTION
**Pull request by TongLi3701**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 1]
```diff
@@ -213,70 +206,59 @@ def step(self, step_idx: int, pbar: Any, **kwargs) -> Optional[float]:
...
-        effective_tokens_count = torch.sum(action_mask, dim=-1) * loss_mask
-        effective_samples = all_reduce_sum(torch.sum(loss_mask), self.plugin)
-        total_effective_tokens_count = all_reduce_sum(torch.sum(effective_tokens_count), self.plugin)
-        total_samples = all_reduce_sum(torch.sum(torch.ones_like(loss_mask, device=loss_mask.device)), self.plugin)
-        self.effective_sample_count += effective_samples.item()
-        self.total_sample_count += total_samples.item()
+        prompt_level_mask = loss_mask.view(self.minibatch_size, self.num_generations)
+
+        # [minibatch_size] -> calculate the number of effective prompts
+        effective_prompts_mask = prompt_level_mask.any(dim=1)
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/grpo_consumer.py:225]: So, here. If any sample in a group is over-length, the result for the whole group will not be used

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 2]
```diff
@@ -213,70 +206,59 @@ def step(self, step_idx: int, pbar: Any, **kwargs) -> Optional[float]:
...
-                }
-                action_mask = action_mask[
-                    : -num_excessive_samples if num_excessive_samples != 0 else action_mask.size(0)
-                ]
-                loss_mask = loss_mask[: -num_excessive_samples if num_excessive_samples != 0 else loss_mask.size(0)]
-                advantages = advantages[: -num_excessive_samples if num_excessive_samples != 0 else advantages.size(0)]
-            else:
-                num_excessive_samples = 0
+            need_update = self.effective_prompt_count >= self.batch_size * self.dp_size
+            excessive_prompts = self.effective_prompt_count - self.batch_size * self.dp_size
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/grpo_consumer.py:234]: here excessive_prompts is the total number of excessive prompts on all ranks, while true_indices is calculated base on the data on a single rank, which is not correct

File: applications/ColossalChat/coati/distributed/grpo_consumer.py (hunk 1 above)
Comment (YeAnbang) [reply applications/ColossalChat/coati/distributed/grpo_consumer.py:225]: image
as we discussed, the overlong samples are only masked out during the competition of loss, not affecting other valid samples within a group

File: applications/ColossalChat/coati/distributed/consumer.py [hunk 3]
```diff
@@ -125,20 +130,20 @@ def loop(self) -> None:
                             batches = self.buffer[
                                 self.dp_rank * self.minibatch_size : (self.dp_rank + 1) * self.minibatch_size
                             ]
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/consumer.py:132]: pad batch here is necessary when `imbs` is smaller than `tMbs`, samples may have differ in size, need to pad before stacking

**Review by YeAnbang — COMMENTED** [review]
Overall looks nice, please view my comments

File: applications/ColossalChat/coati/distributed/consumer.py (hunk 3 above)
Comment (TongLi3701) [reply applications/ColossalChat/coati/distributed/consumer.py:132]: I have already updated the producer which pad samples at first, so different batches have the same size.

File: applications/ColossalChat/coati/distributed/grpo_consumer.py (hunk 2 above)
Comment (TongLi3701) [reply applications/ColossalChat/coati/distributed/grpo_consumer.py:234]: fixed.

**Review by YeAnbang — APPROVED** [review]
